syntax = "proto3";

package serverchat;


service ServerChat {
    rpc connect(ConnectionRequest) returns (ConnectionResponse) {}
    rpc disconnect(DisconnectionRequest) returns (Status) {}
    rpc send_message(Message) returns (Status) {}
    rpc get_chat_stream(Empty) returns (stream Action) {}
}

message Empty {

}

message ConnectionRequest {
    string username = 1;
}

message DisconnectionRequest {
    string user_token = 1;
}

message ConnectionResponse {
    string user_token = 1;
    Status status = 2;
}

message Status {
    bool is_ok = 1;
    string error_message = 2;
}

message Message {
    string user_token = 1;
    string text = 2;
}

message Action {
    enum ActionType {
        CONNECT = 0;
        DISCONNECT = 1;
        SEND_MESSAGE = 2;
    }
    string username = 1;
    ActionType action_type = 2;
    string payload = 3;
}

